<a href="https://colab.research.google.com/github/JesseLynch37/Math152/blob/main/Exploration1_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exploration 1 - Group 2
## Collatz Conjecture
Tanner Boyea, Laura Daniels, Jesse Lynch, Gavin Moss, Alexei Pelyushenko

##About this exploration:
Consider any positive integer *n*. If it is even, divide it by two. If it is odd, replace *n* with 3*n* + 1. Then repeat. The Collatz Conjecture claims that if you start with any positive integer, this process will lead to a repeating cycle of 1, 2, 4. This exploration will explore several different aspects of this process.

The different sections of this exploration include:

*   Verify that all positive integers up to 1,000,000 terminate
*   Number of steps to termination
*   What happens to negative integers
*   Cycle size in negative integers
*   Changing the rules



#Verifying Collatz Conjecture up to 1,000,000:

In [ ]:
#  Importing time to be used to test efficieny of code
from time import time

We know that the numbers 1, 2, and 4 form an infinate cycle.

Therefore, if n is ever 1, 2, or 4, we can stop the loop.

In [ ]:
#  Confirming that first 1,000,000 integers terminate:
time0 = time()

for N in range(1, 1000001):
    n = N
    while n not in [1, 2, 4]:
        if n%2 == 0:
            n //= 2
        else:
            n = 3 * n + 1

print("First 1,000,000 integers successfully terminated in {} seconds.".format(time() - time0))

First 1,000,000 integers successfully terminated in 32.243863582611084 seconds.


If we know all integers smaller than k terminate, we can stop the loop if n ever becomes smaller than k.

If n is even and we know all k < n terminate, we can conclude n terminates without testing it.

More efficient way of testing first 1,000,000 integers:

In [ ]:
#  More efficient way of testing first 1,000,000 integers:
time0 = time()

for N in range(3, 1000001, 2):
    n = N
    while n >= N:
        if n%2 == 0:
            n //= 2
        else:
            n = 3 * n + 1
print("First 1,000,000 integers successfully terminated in {} seconds.".format(time() - time0))

First 1,000,000 integers successfully terminated in 1.2860963344573975 seconds.


We can define a function that replaces a chunk of this code to make future coding easier.

In [2]:
#  Defining a function that makes future code simpler:
def collatz(n):
    return [3*n+1, n//2][n%2 == 0]
#  If n is odd, "n%2 == 0" will be 0, so the function will return 3*n+1.
#  If n is even, "n%2 == 0" will be 1, so the function will return n//2.

In [ ]:
#  More efficient test now using function:
time0 = time()
for N in range(3, 1000001, 2):
    n = N
    while n >= N:
        n = collatz(n)
print("First 1,000,000 integers successfully terminated in {} seconds.".format(time() - time0))

First 1,000,000 integers successfully terminated in 1.9930322170257568 seconds.


This way isn't quite as fast, but still pretty good and much easier to work with.

# Number of steps to terminate:

In [ ]:
#  Number of steps for n to reach 1, 2, or 4 for all positive n up to 100:
for N in range(1, 101):
  n = N
  steps = 0
  while n not in [1, 2, 4]:
    steps += 1
    n = collatz(n)
  print(N, steps)

Of course, the numbers 1, 2, and 4 take 0 steps.

Powers of 2 take k steps, where k satisfies 2^k = n/4.

The number that takes the most steps is 97, with 116 steps.

In [ ]:
#  Finding greatest number of steps for n to reach 1, 2, or 4 for all positive n up to 1,000,000:
max_steps = 0
for N in range(1, 1000001):
  n = N
  steps = 0
  while n not in [1, 2, 4]:
    steps += 1
    n = collatz(n)
  if steps > max_steps:
    print(N, steps)
    max_steps = steps

The number that takes the most steps for all n up to 1,000,000 is 837799, with 522 steps.

In [ ]:
#  Number of steps for n to become smaller than it was initially for all positive n up to 100:
#  Skipping even numbers because number of steps will always be 1.
for N in range(3, 101, 2):
  n = N
  steps = 0
  while n >= N:
    steps += 1
    n = collatz(n)
  print(N, steps)

Every number thats 1 more than a multiple for 4 takes 3 steps.

11 takes 8 steps to get smaller. 12 more than that also takes 8 steps. 20 more than that also takes 8 steps. 12 more than that also takes 8 steps. This +12, +20 pattern seems to continue indefinitely.



In [ ]:
#  Finding greatest number of steps for n to reach get smaller for all positive n up to 1,000,000:
max_steps = 0
for N in range(3, 1000001, 2):
  n = N
  steps = 0
  while n >= N:
    steps += 1
    n = collatz(n)
  if steps > max_steps:
    print(N, steps)
    max_steps = steps

3 6
7 11
27 96
703 132
10087 171
35655 220
270271 267
362343 269
381727 282
626331 287


The number that takes the most steps for all n up to 1,000,000 is 626331, with 287 steps.

# Negative integers:

In [ ]:
#  Testing -1:
n = -1
for i in range(10):
  print(n)
  n = collatz(n)

-1
-2
-1
-2
-1
-2
-1
-2
-1
-2


Starting at -1 leads to a cycle of -1, -2.

This means -2 will also have this result.

In [ ]:
#  Testing -3:
n = -3
for i in range(10):
  print(n)
  n = collatz(n)

-3
-8
-4
-2
-1
-2
-1
-2
-1
-2


Starting at -3 also leads to a -1, -2 cycle.

Conjecture A: all negative integers will lead to a -1, -2 cycle.

In [ ]:
#  Testing Conjecture A for first 100 negative integers:
for N in range(-1, -101, -1):
  n = N
  while n not in [-1, -2]:
    print(N, n)
    n = collatz(n)
print("First 100 negative integers successfully terminated")

Conjecture A failed at -5, which lead to a -5, -14, -7, -20, -10 cycle.

Conjecture B: all negative integers will lead to a -1, -2 cycle or a -5, -14, -7, -20, -10 cycle.

In [ ]:
#  Testing Conjecture B for first 100 negative integers:
for N in range(-1, -101, -1):
  n = N
  while n not in [-1, -2, -5, -14, -7, -20, -10]:
    print(N, n)
    n = collatz(n)
print("First 100 negative integers successfully terminated")

Conjecture B failed at -17, which lead to a cycle with length 18.

Conjecture C: all negative integers will lead to a repeating cycle.


In [ ]:
#  Testing Conjecture C for first 100 negative integers:
for N in range(-1, -10001, -1):
  n = N
  current_chain = []
  while n not in current_chain:
    current_chain.append(n)
    n = collatz(n)
print("First 10,000 negative integers successfully terminated")

First 10,000 negative integers successfully terminated


In [19]:
#  More testing on the terminal properties of large negative numbers:
for N in range(-10000000001, -10000000101, -1):
  n = N
  current_chain = []
  while n not in current_chain:
    current_chain.append(n)
    n = collatz(n)
  loop_len = len(current_chain) - current_chain.index(n)
  print(N, loop_len, current_chain[-loop_len:])

-10000000001 2 [-2, -1]
-10000000002 2 [-2, -1]
-10000000003 2 [-2, -1]
-10000000004 2 [-2, -1]
-10000000005 2 [-2, -1]
-10000000006 2 [-2, -1]
-10000000007 2 [-2, -1]
-10000000008 2 [-2, -1]
-10000000009 2 [-2, -1]
-10000000010 2 [-2, -1]
-10000000011 2 [-2, -1]
-10000000012 2 [-2, -1]
-10000000013 2 [-2, -1]
-10000000014 2 [-2, -1]
-10000000015 2 [-2, -1]
-10000000016 2 [-2, -1]
-10000000017 18 [-68, -34, -17, -50, -25, -74, -37, -110, -55, -164, -82, -41, -122, -61, -182, -91, -272, -136]
-10000000018 5 [-14, -7, -20, -10, -5]
-10000000019 5 [-14, -7, -20, -10, -5]
-10000000020 5 [-14, -7, -20, -10, -5]
-10000000021 2 [-2, -1]
-10000000022 2 [-2, -1]
-10000000023 18 [-68, -34, -17, -50, -25, -74, -37, -110, -55, -164, -82, -41, -122, -61, -182, -91, -272, -136]
-10000000024 2 [-2, -1]
-10000000025 5 [-14, -7, -20, -10, -5]
-10000000026 2 [-2, -1]
-10000000027 2 [-2, -1]
-10000000028 2 [-2, -1]
-10000000029 2 [-2, -1]
-10000000030 2 [-2, -1]
-10000000031 2 [-2, -1]
-10000000032 2 [-2

After testing all negative numbers up to -10,000,000,000 for their final cycle length and final cycle, it was found that all negative numbers end up in one of the following cycles:


1.   [-2, -1]
2.   [-14, -7, -20, -10, -5]
3.   [-74, -37, -110, -55, -164, -82, -41, -122, -61, -182, -91, -272, -136, -68, -34, -17, -50, -25]

Conjecture D: All negative numbers will end up in one of the above cycles.